Update bash ex-scripts to be linter compliant

WalterKolczynski-NOAA opened the pull request:
# Description
Updates the ex-scripts to be compliant with both `shellcheck` and `shfmt`. This does _not_ necessarily enforce some standards not checked by `shellcheck`. Export statements are added to some variables to avoid 'not used' errors. In the future, a more careful analysis should be done to see if these variables are used downstream by executables or if they are truly unused, in which case they should be removed.

Scripts employing MPMD are also updated to use `run_mpmd.sh`. This resulted in some retooling of those scripts.

Refs: #397

# Type of change
- [ ] Bug fix (fixes something broken)
- [ ] New feature (adds functionality)
- [x] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
- Full test suite on Ursa

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [x] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA committed:
Update exscripts for shellcheck except unused variables [SC2034]

WalterKolczynski-NOAA committed:
Apply shfmt to exscripts

WalterKolczynski-NOAA committed:
Remove unneeded APRUNCFP in anlcalc

WalterKolczynski-NOAA committed:
Export "unused" variables

Exports variables reported by shellcheck as being unused. A more careful
examination is needed to remove unnecessary variables.

WalterKolczynski-NOAA committed:
Fix errors introduced during formatting updates

WalterKolczynski-NOAA committed:
Fix error in skew-t script

WalterKolczynski-NOAA committed:
Shfmt skew-T

JessicaMeixner-NOAA commented on scripts/exgfs_wave_post_pnt.sh:
```diff
@@ -77,139 +63,143 @@ printf "\nPreparing input files :\n-------------------------\n"
 mv buoy_log.tmp buoy_log.dat
 
-Nb=$(wc buoy_log.dat | awk '{ print $1 }')
+Nb=$(wc -l < buoy_log.dat)
```
Was it confirmed that all point output is the same for waves after this change?

WalterKolczynski-NOAA replied:
I just compared the contents of `gfs.t12z.bull.tar`, and the contents are identical. Is that sufficient, or do you want me to check all the tarballs?

JessicaMeixner-NOAA replied:
Can we do a quick check on one of the boundary point tar balls as well? That should then be sufficient.

WalterKolczynski-NOAA replied:
Yup, looks good

JessicaMeixner-NOAA replied:
thanks for checking.

JessicaMeixner-NOAA commented on scripts/exgfs_wave_prdgen_bulls.sh:
```diff
@@ -22,125 +22,117 @@
-     err_exit "MISSING BULLETIN INFO"
-   fi
+echo '   Looping over buoys ...'
+echo
```
What's wrong with \n?

WalterKolczynski-NOAA replied:
I didn't make this change manually. Must be something `shfmt` does.

DavidHuber-NOAA replied:
@WalterKolczynski-NOAA I think that this comes from the change you made to .editorconfig:
https://github.com/NOAA-EMC/global-workflow/blame/5fa67e3339e300c9d380e32c98dba541948fbfcc/.editorconfig#L25-L38

DavidHuber-NOAA replied:
Disregard, replied to the wrong comment.

JessicaMeixner-NOAA reviewed (commented):
Can someone point me to where 4 spaces instead of 2 is the standard for the g-w or some linter check?   I was unaware that was a thing, since 2 spaces is very widely used across the g-w. 

This PR needs to be tested on WCOSS2 to ensure that all wave AWIPS, etc jobs are still running as expected as this changes those scripts and I think most of the changes were spaces, but I'm still going through to confirm as github isn't showing all of them as such.  Additionally, wave jobs should be checked that the contents is the same before and after this PR as there were a few places where commands were changed.

aerorahul commented on scripts/exgdas_atmos_chgres_forenkf.sh:
```diff
@@ -73,111 +74,80 @@ DOHYBVAR=${DOHYBVAR:-"NO"}
-LEVS_ENKF=${LEVS_ENKF:-$($NCLEN $ATMFCST_ENSRES pfull)} # get LATB_ENFK
+LONB_ENKF=${LONB_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" grid_xt)} # get LONB_ENKF
+LATB_ENKF=${LATB_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" grid_yt)} # get LATB_ENFK
+LEVS_ENKF=${LEVS_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" pfull)}   # get LATB_ENFK
```
```suggestion
LONB_ENKF=${LONB_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" grid_xt)} # get LONB_ENKF
LATB_ENKF=${LATB_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" grid_yt)} # get LATB_ENKF
LEVS_ENKF=${LEVS_ENKF:-$(${NCLEN} "${ATMFCST_ENSRES}" pfull)}   # get LATB_ENKF
```

DavidHuber-NOAA commented:
> Can someone point me to where 4 spaces instead of 2 is the standard for the g-w or some linter check? I was unaware that was a thing, since 2 spaces is very widely used across the g-w.
> 

@WalterKolczynski-NOAA I think that this comes from the change you made to .editorconfig:
https://github.com/NOAA-EMC/global-workflow/blame/5fa67e3339e300c9d380e32c98dba541948fbfcc/.editorconfig#L25-L38

JessicaMeixner-NOAA commented:
> > Can someone point me to where 4 spaces instead of 2 is the standard for the g-w or some linter check? I was unaware that was a thing, since 2 spaces is very widely used across the g-w.
> 
> @WalterKolczynski-NOAA I think that this comes from the change you made to .editorconfig: https://github.com/NOAA-EMC/global-workflow/blame/5fa67e3339e300c9d380e32c98dba541948fbfcc/.editorconfig#L25-L38

So this is a new thing.   This even seems inconsistent in .editorconfig as well.  Is .editorconfig where we should be looking for these sorts of standards in the future?

WalterKolczynski-NOAA commented:
> > > Can someone point me to where 4 spaces instead of 2 is the standard for the g-w or some linter check? I was unaware that was a thing, since 2 spaces is very widely used across the g-w.
> > 
> > 
> > @WalterKolczynski-NOAA I think that this comes from the change you made to .editorconfig: https://github.com/NOAA-EMC/global-workflow/blame/5fa67e3339e300c9d380e32c98dba541948fbfcc/.editorconfig#L25-L38
> 
> So this is a new thing. This even seems inconsistent in .editorconfig as well. Is .editorconfig where we should be looking for these sorts of standards in the future?

Yes. I will note before this point, no standard was enforced for bash scripts, which is why different scripts had different indentation. This PR is part of applying a standard to all the legacy scripts so enforcement can begin. Four spaces is a compromise between space usage and visibility, and is consistent with the python scripts.

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge develop' into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Apply shfmt to analysis changes

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge origin/develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

DavidHuber-NOAA commented on scripts/exgdas_atmos_chgres_forenkf.sh:
```diff
@@ -73,111 +74,80 @@ DOHYBVAR=${DOHYBVAR:-"NO"}
 USE_CFP=${USE_CFP:-"NO"}
 CFP_MP=${CFP_MP:-"NO"}
-nm=""
-if [ $CFP_MP = "YES" ]; then
-    nm=0
-fi
```
Is this set at the JJob level?

WalterKolczynski-NOAA replied:
It's not used anymore with the switch to using `run_mpmd.sh`.

DavidHuber-NOAA commented on scripts/exgdas_atmos_verfozn.sh:
```diff
@@ -9,42 +9,35 @@
 
-   cpreq "${oznstat}" "./oznstat.${PDY}${cyc}"
+    netcdf=0
+    for filenc4 in diag*nc4.gz; do
+        netcdf=1
+        file=$(echo "${filenc4}" | cut -d'.' -f1-2).gz
+        mv "${filenc4}" "${file}"
+    done
```
This needs a file existence check.  If there are no `diag*nc4.gz` files, then `filenc4` will be assigned the literal `diag*nc4.gz`.

DavidHuber-NOAA replied:
Actually, I think this should be reverted to the previous logic.  There are some variables that could be removed (`data_available`, for instance), but otherwise I don't think this script should change logic.

WalterKolczynski-NOAA replied:
Shellcheck doesn't like looping over `ls`: https://www.shellcheck.net/wiki/SC2045

Recommend we turn on `shopt -s nullglob` (as noted on that page) anyway for all our code to prevent `*` being treated as literal anywhere.

DavidHuber-NOAA replied:
@WalterKolczynski-NOAA Can you add in the `shopt -s nullglob` option here?

aerorahul commented on scripts/exglobal_atmos_ensstat.sh:
```diff
@@ -16,5 +16,5 @@ fi
 export err=$?
 if [[ ${err} -ne 0 ]]; then
-   err_exit "One ore more MPMD jobs failed to calculate ensemble statistics!"
+    err_exit "One ore more MPMD jobs failed to calculate ensemble statistics!"
```
```suggestion
    err_exit "One or more MPMD jobs failed to calculate ensemble statistics!"
```

WalterKolczynski-NOAA replied:
Done.

aerorahul commented on scripts/exglobal_atmos_ensstat.sh:
```diff
@@ -2,8 +2,8 @@
-if [[ -a mpmd_script ]]; then
-   rm -Rf mpmd_script
+if [[ -e mpmd_script ]]; then
+    rm -Rf mpmd_script
```
```suggestion
    rm -f mpmd_script
```

WalterKolczynski-NOAA replied:
Done.

aerorahul commented on scripts/exgfs_prdgen_manager.sh:
```diff
@@ -6,67 +6,40 @@
+    master_file="${COMIN_ATMOS_MASTER}/gfs.${cycle}.master.grb2f${fhr3}"
+    if ! wait_for_file "${master_file}" "${sleep_interval}" "${max_tries}"; then
+        msg="FATAL ERROR: After 2 hours of waiting for GFS POST hour ${fhr3}."
+        err_exit "${msg}"
```
```suggestion
        export err=1
        err_exit "After 2 hours of waiting for GFS POST hour ${fhr3}."
```

WalterKolczynski-NOAA replied:
Done

aerorahul reviewed (commented):
looks ok to me.  just a few comments.
thanks for this.  Next, we need a github action that prevents code from getting in that does not conform to the standards spelled out in `editorconfig`.

WalterKolczynski-NOAA committed:
Make minor updates based on PR review

WalterKolczynski-NOAA committed:
Make missed correction based on review comment

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Update j-jobs that have fallen out of compliance

DavidHuber-NOAA commented on scripts/exglobal_atmos_sfcanl.sh:
```diff
@@ -22,71 +22,70 @@
-NTHREADS_CYCLE=${NTHREADS_CYCLE:-24}
-APRUN_CYCLE=${APRUN_CYCLE:-${APRUN:-""}}
+CYCLESH="${CYCLESH:-${USHgfs}/global_cycle.sh}"
+REGRIDSH="${REGRIDSH:-${USHgfs}/regrid_gsiSfcIncr_to_tile.sh}"
```
This seems like an inconsistency.  Sometimes, when a variables is defined with a fallback option, the fallback is quoted (e.g. https://github.com/NOAA-EMC/global-workflow/pull/4151/files#diff-4cc7890265f85821b79705d4a49753b8bfbdcdd1003129c2a1052ef5ee07542aR49).  Other times (like here), the entire RHS is quoted.  Is one approach preferred?

WalterKolczynski-NOAA replied:
I moved them inside (when necessary) here. I started doing it outside, then stopped because shellcheck doesn't flag it anyway.

DavidHuber-NOAA commented on scripts/exglobal_atmos_tropcy_qc_reloc.sh:
```diff
@@ -11,157 +11,154 @@
 
 # Make sure we are in the $DATA directory
-cd "${DATA}"
+cd "${DATA}" || exit
```
Should probably be `exit 1` so the J-Job handles the failure correctly.

WalterKolczynski-NOAA replied:
Done.

DavidHuber-NOAA commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -29,254 +29,245 @@ fi
-       last=${ncount}
+    # Each set represents a group of files
+    if [[ "${nset}" == 1 ]]; then
+        grp="a" # TODO: this should be "a" when we eventually rename the pressure grib2 files per EE2 convention
```
This work has been completed
```suggestion
        grp="a"
```

WalterKolczynski-NOAA replied:
Fixed

DavidHuber-NOAA commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -29,254 +29,245 @@ fi
-              -grib "${COMOUT_ATMOS_GRIB_0p25}/${PREFIX}wgne.${fhr3}.grib2"
-  fi
+    grp="a"
+    if [[ "${FORECAST_HOUR}" -gt 0 && "${FORECAST_HOUR}" -le ${FHMAX_WGNE:-0} ]]; then
```
Just a note that double-quotes are not needed when doing arithmetic logic.  No need to change.

WalterKolczynski-NOAA replied:
I try to always used them in conditionals, because even though they are not _required_, if the variable is unset you get a "conditional binary operator expected" warning.

DavidHuber-NOAA replied:
Thanks for the pointer!

aerorahul replied:
wouldn't the `set -u` trap it first?

DavidHuber-NOAA commented on scripts/exglobal_atmos_tropcy_qc_reloc.sh:
```diff
@@ -11,157 +11,154 @@
 tmmark_uc=${tmmark^^}
 
-iflag=0
+export iflag=0
```
No need to export this.  `iflag` isn't used upstream.
```suggestion
iflag=0
```

WalterKolczynski-NOAA replied:
If it's not being used downstream, then it isn't being used at all. Removing.

DavidHuber-NOAA commented on scripts/exglobal_diag.sh:
```diff
@@ -47,75 +47,70 @@ rm -f "${RADSTAT}" "${PCPSTAT}" "${CNVSTAT}" "${OZNSTAT}"
+    diagtype[0]="conv conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_tcp conv_uv conv_spd"
+    diagtype[1]="pcp_ssmi_dmsp pcp_tmi_trmm"
+    diagtype[2]="sbuv2_n16 sbuv2_n17 sbuv2_n18 sbuv2_n19 gome_metop-a gome_metop-b omi_aura mls30_aura ompsnp_npp ompstc8_npp  ompstc8_n20 ompsnp_n20 ompstc8_n21 ompsnp_n21 ompslp_npp gome_metop-c"
+    diagtype[3]="msu_n14 sndr_g08 sndr_g11 sndr_g12 sndr_g13 sndr_g08_prep sndr_g11_prep sndr_g12_prep sndr_g13_prep sndrd1_g11 sndrd2_g11 sndrd3_g11 sndrd4_g11 sndrd1_g12 sndrd2_g12 sndrd3_g12 sndrd4_g12 sndrd1_g13 sndrd2_g13 sndrd3_g13 sndrd4_g13 sndrd1_g14 sndrd2_g14 sndrd3_g14 sndrd4_g14 sndrd1_g15 sndrd2_g15 sndrd3_g15 sndrd4_g15 amsua_n15 amsua_n16 amsua_n17 amsub_n15 amsub_n16 amsub_n17 hs ...
```
This was probably duplicated with a merge.

```suggestion
```

WalterKolczynski-NOAA replied:
Yeah, some of these merges were ugly. Fixed.

DavidHuber-NOAA commented on scripts/exglobal_atmos_analysis.sh:
```diff
@@ -336,213 +334,218 @@ NST=${NST:-""}
+${NLN} "${BERROR}" berror_stats
+${NLN} "${SATANGL}" satbias_angle
+if [[ "${SATINFO}" == "generate" ]]; then
+    # shellcheck disable=SC2153
+    "${USHgfs}/create_gsi_info.sh" sat "${PDY}${cyc}" "${DATA}"
```
What variable(s) did shellcheck think were misspelled here?

WalterKolczynski-NOAA replied:
`${DATA}`

DavidHuber-NOAA reviewed (changes requested):
Overall this is a great changeset!  Thanks Walter.  There's just a couple of minor issues that need addressing and then we can start testing.

WalterKolczynski-NOAA committed:
Merge develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Correct issues found in code review

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA The changes look good. There are just a couple of conflicts in the branch. Can you merge in develop and resolve the conflicts? After that, I think this is good to begin testing.

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

DavidHuber-NOAA commented:
Launching tests on all platforms.  @WalterKolczynski-NOAA Can you please perform a 1-to-1 test on one of the platforms as well? I will pass you the paths for the nightly runs in the morning, as well as the paths to the RUNTESTS from this PR, for comparison.

TerrenceMcGuinness-NOAA commented:
[Ursa failed on the nightly](https://vlab.noaa.gov/gitlab-community/NWS/Operations/NCEP/EMC/global-workflow/-/jobs/34202) as well after I restarted the timeout.  Here is error log file [enkfgdas_atmensanlobs.log](https://gist.github.com/emcbot/7852de44e34dd9cb22947cc6baabbd8e#file-enkfgdas_atmensanlobs-log)

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA It looks like the Ursa failure in last night's nightly run was due to a timeout of the `enkfgdas_atmensanlobs` job, while this PR failed due to a potential scripting issue in the `gdas_prepatmiodaobs` job. Walter, could you take a look at the failed job?

aerorahul committed:
Merge branch 'develop' into feature/shfmt_exscripts

TerrenceMcGuinness-NOAA commented:
Gaea C6 Runner was of line this morning and I didn't get it back until this evening (working remotely in other time zone) it is running now.

TerrenceMcGuinness-NOAA commented:
@DavidHuber-NOAA The [rebuild on Gaea C6 ](https://vlab.noaa.gov/gitlab-community/NWS/Operations/NCEP/EMC/global-workflow/-/jobs/34989) for this PR had some inexplicable errors (I have not pursued this yet -- see below) While the [restart from the nightly worked fine](https://vlab.noaa.gov/gitlab-community/NWS/Operations/NCEP/EMC/global-workflow/-/jobs/34988).  We'll see in the morning after the nightly runs again.

```
pipeline is set to checkout PR_NUMBER=4151 from https://github.com/NOAA-EMC/global-workflow on machine Gaeac6
GraphQL: Projects (classic) is being deprecated in favor of the new Projects experience, see: https://github.blog/changelog/2024-05-23-sunset-notice-projects-classic/. (repository.pullRequest.projectCards)
Running after_script
$ echo "After script is running for job ${CI_JOB_NAME} on ${machine} with status ${CI_JOB_STATUS}" # collapsed multi-line command
After script is running for job build-gaeac6 on gaeac6 with status failed
Job failed, performing cleanup actions for Gaeac6
Updating GitHub labels for PR 4151 failure on Gaeac6
GraphQL: Projects (classic) is being deprecated in favor of the new Projects experience, see: https://github.blog/changelog/2024pipeline is set to checkout PR_NUMBER=4151 from https://github.com/NOAA-EMC/global-workflow on machine Gaeac6
GraphQL: Projects (classic) is being deprecated in favor of the new Projects experience, see: https://github.blog/changelog/2024-05-23-sunset-notice-projects-classic/. (repository.pullRequest.projectCards)
Running after_script
00:00
Running after script...
$ echo "After script is running for job ${CI_JOB_NAME} on ${machine} with status ${CI_JOB_STATUS}" # collapsed multi-line command
After script is running for job build-gaeac6 on gaeac6 with status failed
Job failed, performing cleanup actions for Gaeac6
Updating GitHub labels for PR 4151 failure on Gaeac6
GraphQL: Projects (classic) is being -05-23-sunset-notice-projects-classic/. (repository.pullRequest.projectCards)
Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit status 1
```

aerorahul reviewed (dismissed):
Looks good.

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA I recall these GraphQL errors coming up before. It required an updated github-cli IIRC. I'm going to go ahead and launch this on C6 manually.

DavidHuber-NOAA commented:
Ursa passed on the second run through, so I suspect the `prepiodaobs` failure (noted 5 days ago [here](https://github.com/NOAA-EMC/global-workflow/pull/4151#issuecomment-3563045837)) was transient. Once the `shopt` option is added in and the conflicts are resolved, I will merge the changes locally on C6 and start the tests there. Once those pass, we can merge this PR.

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

WalterKolczynski-NOAA commented:
> Ursa passed on the second run through, so I suspect the `prepiodaobs` failure (noted 5 days ago [here](https://github.com/NOAA-EMC/global-workflow/pull/4151#issuecomment-3563045837)) was transient. Once the `shopt` option is added in and the conflicts are resolved, I will merge the changes locally on C6 and start the tests there. Once those pass, we can merge this PR.

What `shopt` option? Is this something I need to do?

DavidHuber-NOAA commented:
> > Ursa passed on the second run through, so I suspect the `prepiodaobs` failure (noted 5 days ago [here](https://github.com/NOAA-EMC/global-workflow/pull/4151#issuecomment-3563045837)) was transient. Once the `shopt` option is added in and the conflicts are resolved, I will merge the changes locally on C6 and start the tests there. Once those pass, we can merge this PR.
> 
> What `shopt` option? Is this something I need to do?

@WalterKolczynski-NOAA Yes, this was your suggestion to an issue I raised: https://github.com/NOAA-EMC/global-workflow/pull/4151#discussion_r2518210180

WalterKolczynski-NOAA committed:
Turn on null globs

Turns on the bash option to allow null globs in loops rather than
treating `*` as a literal.

WalterKolczynski-NOAA commented:
> > > Ursa passed on the second run through, so I suspect the `prepiodaobs` failure (noted 5 days ago [here](https://github.com/NOAA-EMC/global-workflow/pull/4151#issuecomment-3563045837)) was transient. Once the `shopt` option is added in and the conflicts are resolved, I will merge the changes locally on C6 and start the tests there. Once those pass, we can merge this PR.
> > 
> > 
> > What `shopt` option? Is this something I need to do?
> 
> @WalterKolczynski-NOAA Yes, this was your suggestion to an issue I raised: [#4151 (comment)](https://github.com/NOAA-EMC/global-workflow/pull/4151#discussion_r2518210180)

Ah. Find wasn't showing it for some reason. Change made. Put in preamble for want of a better location.

DavidHuber-NOAA commented:
Great, thanks @WalterKolczynski-NOAA!  I'll start up testing on C6.

WalterKolczynski-NOAA committed:
Merge branch develop into feature/shfmt_exscripts

DavidHuber-NOAA commented:
All tests passed. I confirmed there were no changes between this PR and the nightly run for the C96C48mx500_cyc_gfs test case on C6. Merging.
